### PR TITLE
Events Plugin - Add initial "NewBlock" event message

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

* It returns for any block we would broadcast to other peers, when
  we would broadcast them.
* It returns a JSON String containing hask, number, and timestamp
* This event data is not set in stone, it may change in type or content.
* Acceptance tests and unit tests got a re-work away from the assumption
  that there is only one plugin type.

### CODE REVIEW

File: ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/BlockBroadcaster.java [hunk 1]
```diff
@@ -16,21 +16,34 @@
...
+  public long subscribePropagateNewBlocks(final Consumer<Block> callback) {
+    return newBlockSubscribers.subscribe(callback);
+  }
+
+  public void unsubscribePropagateNewBlocks(final long id) {
+    newBlockSubscribers.unsubscribe(id);
+  }
+
   public void propagate(final Block block, final UInt256 totalDifficulty) {
+    newBlockSubscribers.forEach(listener -> listener.accept(block));
```
Comment (mbaxter) [inline ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/BlockBroadcaster.java]: I don't think you want to add this functionality here.  `Blockchain.observeBlockAdded` already supports observation of new block events.

Comment (shemnon) [reply ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/BlockBroadcaster.java]: This event needs to be fired between when we validate the header and when we validate the entire block.

The stack for this code is in part
```
Breakpoint reached
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockBroadcaster.propagate(BlockBroadcaster.java:46)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.broadcastBlock(BlockPropagationManager.java:248)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.lambda$validateAndProcessPendingBlock$6(BlockPropagationManager.java:306)
	  at ...
```

And the stack for using the `Blockchain.observeBlockAdded` event is
```
...
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.runImportTask(BlockPropagationManager.java:323)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.validateAndProcessPendingBlock(BlockPropagationManager.java:307)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.lambda$importOrSavePendingBlock$5(BlockPropagationManager.java:297)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager$$Lambda$1088.740466537.get(Unknown Source:-1)
...
```

Like 307 for BlockPropagationManager runs the full block validation, which is later than the requirements we have for when this information should be pushed to the events plugin.  We don't want full block validation for this event just header (e.g. proof of work) validation.

File: ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/BlockBroadcaster.java [hunk 2]
```diff
@@ -16,21 +16,34 @@
...
+import java.util.function.Consumer;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class BlockBroadcaster {
   private static final Logger LOG = LogManager.getLogger();
 
   private final EthContext ethContext;
+  private final Subscribers<Consumer<Block>> newBlockSubscribers = new Subscribers<>();
```
Comment (mbaxter) [inline ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/BlockBroadcaster.java]:
```suggestion
  private final Subscribers<Consumer<Block>> blockPropagatedSubscribers = new Subscribers<>();
```

File: pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  private static Logger LOG = LogManager.getLogger();
+
+  private final BlockBroadcaster blockBroadcaster;
+
+  public PantheonEventsImpl(final BlockBroadcaster blockBroadcaster) {
+    this.blockBroadcaster = blockBroadcaster;
+  }
+
+  @Override
+  public Object addBlockAddedListener(final Consumer<String> blockJSONListener) {
```
Comment (mbaxter) [inline pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java]:
```suggestion
  public Object observeNewBlockPropagated(final Consumer<String> blockJSONListener) {
```

File: pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    LOG.info("Adding Listener");
+    return blockBroadcaster.subscribePropagateNewBlocks(
+        block -> dispatchNewBlockMessage(block, blockJSONListener));
+  }
+
+  @Override
+  public void removeBlockAddedObserver(final Object listenerIdentifier) {
```
Comment (mbaxter) [inline pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java]:
```suggestion
  public void removeNewBlockPropagatedObserver(final Object listenerIdentifier) {
```

File: pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    LOG.info("Removing Listener");
+    if (listenerIdentifier instanceof Long) {
+      blockBroadcaster.unsubscribePropagateNewBlocks((Long) listenerIdentifier);
+    }
+  }
+
+  private void dispatchNewBlockMessage(final Block block, final Consumer<String> callback) {
```
Comment (mbaxter) [inline pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java]:
```suggestion
  private void dispatchNewBlockPropagatedMessage(final Block block, final Consumer<String> callback) {
```

File: ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/BlockBroadcaster.java (hunk 1 above)
Comment (mbaxter) [reply ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/BlockBroadcaster.java]: Ok, that makes sense.  I think we just need to change the naming to make it clear what this event is.  In the future if we handle NEW_BLOCK_HASHES / NEW_BLOCK messages differently, we may want to expose more events to get all of the new blocks on the network.  Most clients don't propagate out NEW_BLOCK messages in response to receiving NEW_BLOCK_HASHES from the network, for example.

File: pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java (hunk 3 above)
Comment (mbaxter) [inline pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java]: (optional) Consider adding a functional interface in place of `Consumer<String>`.

Comment (shemnon) [reply pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java]: Listener is more prevalent in our code base than observer.  So I added the Propagated but kept the Listener.

File: pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java (hunk 4 above)
Comment (shemnon) [reply pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java]: Listener is more prevalent in our code base than observer.  So I added the Propagated but kept the Listener.

File: pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java (hunk 5 above)
Comment (shemnon) [reply pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java]: Listener is more prevalent in our code base than observer.  So I added the Propagated but kept the Listener.

File: pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java (hunk 3 above)
Comment (shemnon) [reply pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java]: Done.

File: pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+public class PantheonEventsImpl implements PantheonEvents {
+  private final BlockBroadcaster blockBroadcaster;
+
+  public PantheonEventsImpl(final BlockBroadcaster blockBroadcaster) {
+    this.blockBroadcaster = blockBroadcaster;
+  }
+
+  @Override
+  public Object addNewBlockPropagatedLister(final NewBlockPropagatedListener listener) {
```
Comment (mbaxter) [inline pantheon/src/main/java/tech/pegasys/pantheon/services/PantheonEventsImpl.java]:
```suggestion
  public Object addNewBlockPropagatedListener(final NewBlockPropagatedListener listener) {
```
Typo

File: build.gradle [hunk 7]
```diff
@@ -105,7 +105,7 @@ allprojects {
     java {
       // This path needs to be relative to each project
       target fileTree('.') {
-        include '**/*.java'
+        include '**/src/*/java/**/*.java'
```
Comment (mbaxter) [inline build.gradle]: what are these changes for??

Comment (shemnon) [reply build.gradle]: meant for another PR.  Spotless was catching intellij generated files from .../out/classes/....  By limiting src to having to live in src directories generated and transient java files don't get style checked.

**Pull request merged**